File: Data Structures & Algorithms/decode-ways/submission-1.cpp
class Solution {
public:
    // 4 4 4 1 4 5 5 1 1
    int numDecodings(string s) {
        if(s[0]=='0')
            return 0;
        int res=0;
        int n=s.size();
        vector<int> dp(n,0);

        dp[0]=1;
        string temp;
        for(int i=1;i<n;i++){
            temp = "";

            if(s[i]=='0'){
                if(s[i-1]=='1' || s[i-1]== '2')
                    dp[i] = i>1 ? dp[i-2] : 1;
                else return 0;   

                continue; 
            }
            // d d d a a
            // d d d o

            dp[i] = dp[i-1];

            if(s[i-1]=='1'){
                dp[i] += i>1 ?(dp[i-2]):1;
            }else if(s[i-1]=='2' && (s[i] <'7')){
                dp[i] += i>1 ?(dp[i-2]):1;
            }    
            
            

        }

        return dp[n-1];
    }
};

File: Data Structures & Algorithms/reverse-nodes-in-k-group/submission-1.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */

// 1 2 3 4 5 6
// 2 1 4 3
class Solution {
    

    public void reverse(ListNode start, ListNode end) {
    ListNode prev = end.next; // Connect the new tail to the rest of the list
    ListNode temp = start;
    ListNode stop = end.next; // Define exactly where to stop

    while (temp != stop) {
        ListNode next = temp.next;
        temp.next = prev;
        prev = temp;
        temp = next;
    }
    // Note: After this, 'prev' is the new head of this segment.
    // If you need to update the node before 'start', you must do it outside this method.
}
    public ListNode reverseKGroup(ListNode head, int k) {
        
        ListNode res= null;

        ListNode temp=head,start=head,end=null,prevEnd = null,next;

        int ctr=0;
        while(temp!=null){
            ctr ++;
            next = temp.next;

            if(ctr == k){
                end = temp;
                reverse(start,end);
                start.next = next;

                if(res == null){
                    res = end;
                }else {
                    prevEnd.next = end;
                }
                prevEnd = start;
                ctr=0;
                start = next;
                
            }

            temp = next;
        }

        return (res == null) ? head : res;
    }
}
